polynote-server/src/main/scala/polynote/server/package.scala

```diff
@@ -233,6 +235,22 @@ package object server {
           case Some(publisher) => publisher.kernelStatus()
         }
 
+        override def search(query: String): RIO[BaseEnv with GlobalEnv, List[NotebookSearchResult]] = {
+          val validCells = repository.listNotebooks.flatMap(nbs => ZIO.foreachParN(16)(nbs) { nb => {
+            for {
+              loadedNB <- repository.loadNotebook(nb)
+              cells    <- ZIO(loadedNB.cells.filter(c => c.content.toString.contains(query)))
+            } yield (loadedNB.path, cells)
+          }})
+
+          validCells.map(nb => {
+            for {
+              cells <- nb
+              cell   <- cells._2
+            } yield NotebookSearchResult(cells._1, cell.id, cell.content.toString)
+          })
+        }
+
```

Comment:
Feedback is greatly appreciated on how to re-write this function - it works, but I couldn't find a cleaner way other than this weird tuple solution. Just blocked on how to leverage ZIO/Scala expressions more efficiently for this...

Reply:
You could do something like this instead:

```scala
override def search(query: String): RIO[BaseEnv with GlobalEnv, List[NotebookSearchResult]] =
  ZIO.foreachParN(16)(repository.listNotebooks) {
    nb => repository.loadNotebook(nb).flatMap {
      loaded => for {
        cell <- loaded.cells if cell.content.toString contains query
      } yield NotebookSearchResult(loaded.path, cell.id, cell.content.toString)
    }
  }
```